salesagility/SuiteCRM
# Fix #10282 "Campaign Response by Recipient Activity" (Status page) ha…

…s recently become unreliable. And others.



With this I implement  [RFC8085, "Signaling One-Click Functionality for List Email Headers"](https://datatracker.ietf.org/doc/html/rfc8058) to restore the accurate working of the SuiteCRM opt-out feature for Newsletters and Email Campaigns, by preventing email services from automatically opting-out subscribers by crawling links in the email message, which before this Fix, used to load the opt-out link, and perform the opt-out.

All major email services (Microsoft, Gmail, Yahoo, etc) are crawling the links in email messages, to protect users from malware, and that crawling has been triggering the unsubscribe ("opt-out") feature, the "Remove Me" link in the footer of every Campaign and Newsletter bulk email message.

The new way, required in 2024 for all mail sent to Gmail and Yahoo, is to have new headers in the email, telling the email client app exactly which URL to request with a special action that crawlers don't use, to enable the user to do a one-click unsubscribe ("opt-out"), for the subscriber, for this particular email campaign or newsletter.  So the subscriber no longer has to leave their email client app, visit the Remove Me web page on the CRM, or even click a button to unsubscribe.

## Description




1. Improve reliability in detecting the web app's URL, in case the admin has left 'site_url' blank. This is required for constructing the opt-out URL.
2. Fix some missing global statements for `$log` and `$mod_strings`.
3. Accurantly construct the headers required for RFC8085 one-click unsubscribe.

## Motivation and Context

The top Email services have been crawling all the links in the Campaign and Newsletter emails for a long time.  One load of the opt-out (unsubscribe) link was actually unsubscribing every single lead or target who received the email, before they could even open it and read it!

## How To Test This

Without this Fix, try sending a Campaign or Newsletter to only gmail, yahoo, or microsoft hosted email address.  All of your emails will unsubsribe immediately.  That's the automated link scanners loading the opt-out link to check for malware!

Apply this Fix.  Send a Campaign or Newsletter to only Gmail, Yahoo, and Microsoft hosted emil.  The headers for RFC8085 should be present, and none of the recipients will unsubscribe until they have a chance to open up the email message and look at it.  In the email app, the subscriber will see a new "Unsubscribe" button, located near the headers.  Clicking on this new "Unsubscribe" button should perform the opt-out from the SuiteCRM Newsletter or Email Campaign, from inside the email app.

## Types of changes


### Final checklist
